File: python/src/captchakraken/humanize.py
"""How the driver moves: one pluggable object per input device. Mirrors js/src/humanize.ts.

The pointer position lives here, not in the solver, because a touch mode that dispatches no
motion between taps still has to say where the next gesture starts.
"""

from __future__ import annotations

import os
import random
import time
from typing import Any, Dict, List, Optional, Sequence, Tuple

from cursory import generate_trajectory

Point = Tuple[float, float]

# Every inter-gesture wait the driver takes, named, so each device supplies its own table.
PAUSE_KINDS = ("tap", "between", "grab", "drop", "probe", "settle", "key")


def _delay(ms: float) -> None:
    if ms > 0:
        time.sleep(ms / 1000.0)


def _log(message: str) -> None:
    print(f"[captchakraken] {message}", flush=True)


def _same_point(a: Point, b: Point) -> bool:
    return abs(float(a[0]) - float(b[0])) < 1e-6 and abs(float(a[1]) - float(b[1])) < 1e-6


def trajectory(start: Point, end: Point, frequency: int = 60) -> Tuple[List[Point], List[float]]:
    """A recorded human movement morphed onto the endpoints; timings are cumulative ms."""
    points, timings = generate_trajectory(start, end, frequency=frequency)
    return [(float(x), float(y)) for x, y in points], [float(t) for t in timings]


class Humanizer:
    """The gesture vocabulary the driver speaks. Subclass or duck-type."""

    name = "custom"
    hovers = False
    PAUSES: Dict[str, Tuple[float, float]] = {}

    def __init__(self, start: Point = (0.0, 0.0)) -> None:
        self.at: Point = (float(start[0]), float(start[1]))

    def reset(self, page: Any) -> None:
        pass

    def move(self, page: Any, to: Point) -> None:
        raise NotImplementedError

    def press(self, page: Any) -> None:
        raise NotImplementedError

    def release(self, page: Any) -> None:
        raise NotImplementedError

    def type_text(self, page: Any, field: Any, text: str) -> bool:
        raise NotImplementedError

    def _pause_ms(self, kind: str) -> float:
        lo, hi = self.PAUSES.get(kind, (0.0, 0.0))
        return random.uniform(lo, hi)

    def pause(self, kind: str) -> None:
        _delay(self._pause_ms(kind))

    def click(self, page: Any, to: Point) -> None:
        self.move(page, to)
        self.press(page)
        self.pause("tap")
        self.release(page)

    def drag(self, page: Any, src: Point, dst: Point) -> None:
        self.move(page, src)
        self.press(page)
        self.pause("grab")
        self.move(page, dst)
        self.pause("drop")
        self.release(page)


class MouseHumanizer(Humanizer):
    name = "mouse"
    hovers = True
    PAUSES = {"tap": (20.0, 50.0), "between": (80.0, 160.0), "grab": (50.0, 100.0), "drop": (50.0, 100.0),
              "probe": (40.0, 80.0), "settle": (90.0, 210.0), "key": (45.0, 135.0)}

    def __init__(self, start: Point = (0.0, 0.0), frequency: int = 60) -> None:
        super().__init__(start)
        self._frequency = frequency
        self._viewport_cache: Optional[Dict[str, float]] = None
        self._down = False
        self._cursor_seeded = False

    def reset(self, page: Any) -> None:
        self._viewport_cache = None
        self._cursor_seeded = False
        self._down = False

    def _viewport(self, page: Any) -> Optional[Dict[str, float]]:
        """The window to keep the cursor inside; None under camoufox, which reports no viewport."""
        if self._viewport_cache is None:
            for read in (lambda: page.viewport_size() if callable(page.viewport_size) else page.viewport_size,
                         lambda: page.evaluate("() => ({width: window.innerWidth, height: window.innerHeight})")):
                try:
                    vp = read()
                    if vp and vp.get("width") and vp.get("height"):
                        self._viewport_cache = {"width": float(vp["width"]), "height": float(vp["height"])}
                        break
                except Exception:
                    continue
        return self._viewport_cache

    def _seed_cursor(self, page: Any) -> None:
        """One plain move off the (0, 0) origin per solve: a trajectory from the corner wedges camoufox."""
        if self._cursor_seeded:
            return
        self._cursor_seeded = True
        if self.at != (0.0, 0.0):
            cx, cy = self.at
        else:
            vp = self._viewport(page)
            cx, cy = (vp["width"] / 2, vp["height"] / 2) if vp else (200.0, 200.0)
        try:
            page.mouse.move(cx, cy)
            self.at = (cx, cy)
        except Exception:
            pass

    def move(self, page: Any, to: Point) -> None:
        self._seed_cursor(page)
        if _same_point(self.at, to):
            return
        points, timings = trajectory(self.at, to, self._frequency)
        self._trace(page, points, timings)

    def _trace(self, page: Any, points: Sequence[Point], timings: Sequence[float]) -> None:
        # Clamp only when the viewport is known: a guessed edge coordinate deadlocks camoufox's juggler.
        viewport = self._viewport(page)
        start = time.monotonic() * 1000.0
        for (x, y), t in zip(points, timings):
            x, y = float(x), float(y)
            if viewport is not None:
                x = max(1.0, min(x, viewport["width"] - 1.0))
                y = max(1.0, min(y, viewport["height"] - 1.0))
            try:
                page.mouse.move(x, y)
            except Exception as exc:
                if "Target closed" in str(exc) or "Session closed" in str(exc):
                    _log("could not move mouse; page or session closed")
                    return
                continue
            self.at = (x, y)
            _delay(start + t - time.monotonic() * 1000.0)

    def press(self, page: Any) -> None:
        page.mouse.down()
        self._down = True

    def release(self, page: Any) -> None:
        page.mouse.up()
        self._down = False

    def type_text(self, page: Any, field: Any, text: str) -> bool:
        try:
            page.keyboard.press("Control+A")
        except Exception:
            pass
        for ch in text:
            try:
                page.keyboard.type(ch)
            except Exception as exc:
                _log(f"could not type into the captcha field: {exc}")
                return False
            self.pause("key")
        return True


class TouchBackend:
    """Where touch events go. `move` takes a whole leg of `(x, y, dt_ms)` samples."""

    name = "touch"

    def down(self, x: float, y: float) -> None:
        raise NotImplementedError

    def move(self, path: Sequence[Tuple[float, float, float]]) -> None:
        raise NotImplementedError

    def up(self, x: float, y: float) -> None:
        raise NotImplementedError


class CdpTouchBackend(TouchBackend):
    """`Input.dispatchTouchEvent` over CDP; Chromium-family pages launched with has_touch=True."""

    name = "cdp"

    def __init__(self, page: Any) -> None:
        try:
            context = page.context() if callable(page.context) else page.context
            self._session = context.new_cdp_session(page)
        except Exception as exc:
            raise RuntimeError(
                f"mobile humanisation needs touch dispatch, and this page offers none ({exc}). Use a "
                "Chromium-family Playwright browser launched with has_touch=True, or pass an Appium "
                "driver as PageSolverConfig.touch_driver.") from exc

    def _send(self, kind: str, points: Sequence[Tuple[float, float]]) -> None:
        # radius and force are what a real digitizer reports; a synthetic tap reports zero.
        self._session.send("Input.dispatchTouchEvent", {"type": kind, "touchPoints": [
            {"x": float(x), "y": float(y), "radiusX": random.uniform(8.0, 14.0),
             "radiusY": random.uniform(8.0, 14.0), "force": random.uniform(0.35, 0.75), "id": 1}
            for x, y in points]})

    def down(self, x: float, y: float) -> None:
        self._send("touchStart", [(x, y)])

    def move(self, path: Sequence[Tuple[float, float, float]]) -> None:
        for x, y, dt_ms in path:
            _delay(dt_ms)
            self._send("touchMove", [(x, y)])

    def up(self, x: float, y: float) -> None:
        self._send("touchEnd", [])


class AppiumTouchBackend(TouchBackend):
    """W3C touch pointer actions for Appium and Selenium, paced by the device from one chain per leg."""

    name = "appium"

    def __init__(self, driver: Any, scale: Optional[float] = None, origin: Optional[Point] = None,
                 page: Any = None) -> None:
        self._driver = driver
        self._scale_given = scale is not None
        self._scale = 1.0 if scale is None else float(scale)
        self._origin = (0.0, 0.0) if origin is None else (float(origin[0]), float(origin[1]))
        self._page = page
        self._checked = False

    def _check_scale(self) -> None:
        """Refuse an unset scale on a device whose pixel ratio is not 1: the finger would land elsewhere."""
        self._checked = True
        if self._scale_given or self._page is None or not hasattr(self._page, "evaluate"):
            return
        try:
            dpr = float(self._page.evaluate("() => window.devicePixelRatio"))
        except Exception:
            return
        if abs(dpr - 1.0) < 1e-6:
            return
        raise RuntimeError(
            f"the touch driver maps CSS pixels onto a device reporting devicePixelRatio {dpr:g}, and no "
            f"scale was given. Fix: touch_transform={{'scale': {dpr:g}, 'origin': (x, y)}}, where origin "
            "is the webview's top-left in SCREEN coordinates, or pass scale=1 to assert the coordinates "
            "are already mapped.")

    def _map(self, x: float, y: float) -> Tuple[int, int]:
        if not self._checked:
            self._check_scale()
        return int(round(self._origin[0] + x * self._scale)), int(round(self._origin[1] + y * self._scale))

    def _perform(self, actions: Sequence[Dict[str, Any]]) -> None:
        chain = [{"type": "pointer", "id": "ck-finger", "parameters": {"pointerType": "touch"},
                  "actions": list(actions)}]
        if callable(getattr(self._driver, "execute_actions", None)):
            self._driver.execute_actions(chain)
        elif callable(getattr(self._driver, "execute", None)):
            self._driver.execute("actions", {"actions": chain})
        else:
            raise RuntimeError("the touch driver speaks neither execute_actions() nor execute('actions', …); "
                               "pass a WebDriver-compatible driver or a custom TouchBackend.")

    def down(self, x: float, y: float) -> None:
        mx, my = self._map(x, y)
        self._perform([{"type": "pointerMove", "duration": 0, "origin": "viewport", "x": mx, "y": my},
                       {"type": "pointerDown", "button": 0}])

    def move(self, path: Sequence[Tuple[float, float, float]]) -> None:
        actions = []
        for x, y, dt_ms in path:
            mx, my = self._map(x, y)
            actions.append({"type": "pointerMove", "duration": max(0, int(round(dt_ms))),
                            "origin": "viewport", "x": mx, "y": my})
        if actions:
            self._perform(actions)

    def up(self, x: float, y: float) -> None:
        if not self._checked:
            self._check_scale()
        self._perform([{"type": "pointerUp", "button": 0}])


class TouchscreenTouchBackend(TouchBackend):
    """Playwright's tap-only touchscreen, for browsers with touch but no CDP. A drag cannot be expressed."""

    name = "touchscreen"

    def __init__(self, page: Any) -> None:
        self._page = page
        self._pending: Optional[Point] = None

    def down(self, x: float, y: float) -> None:
        self._pending = (x, y)

    def move(self, path: Sequence[Tuple[float, float, float]]) -> None:
        raise RuntimeError("this browser exposes taps but not touch travel, so a drag/slide puzzle cannot be "
                           "driven on it. Use a Chromium-family browser (CDP touch dispatch) or an Appium driver.")

    def up(self, x: float, y: float) -> None:
        at = self._pending or (x, y)
        self._pending = None
        self._page.touchscreen.tap(float(at[0]), float(at[1]))


def touch_backend_for(page: Any, driver: Any = None, **kwargs: Any) -> TouchBackend:
    if driver is not None:
        return driver if isinstance(driver, TouchBackend) else AppiumTouchBackend(driver, page=page, **kwargs)
    try:
        return CdpTouchBackend(page)
    except RuntimeError:
        if hasattr(page, "touchscreen"):
            _log("no CDP session; falling back to tap-only touch dispatch")
            return TouchscreenTouchBackend(page)
        raise


class MobileHumanizer(Humanizer):
    """A finger on glass: no hover, touch events only, slower and more variable pauses."""

    name = "mobile"
    hovers = False
    PAUSES = {"tap": (55.0, 130.0), "between": (140.0, 320.0), "grab": (90.0, 190.0), "drop": (80.0, 170.0),
              "probe": (70.0, 140.0), "settle": (140.0, 300.0), "key": (110.0, 320.0)}

    def __init__(self, start: Point = (0.0, 0.0), backend: Any = None, driver: Any = None,
                 frequency: int = 90, **backend_kwargs: Any) -> None:
        super().__init__(start)
        self._backend: Optional[TouchBackend] = backend
        self._driver = driver
        self._backend_kwargs = backend_kwargs
        self._frequency = frequency
        self._down = False

    def reset(self, page: Any) -> None:
        if self._down:
            try:
                self._touch(page).up(*self.at)
            except Exception:
                pass
            self._down = False

    def _touch(self, page: Any) -> TouchBackend:
        if self._backend is None:
            self._backend = touch_backend_for(page, self._driver, **self._backend_kwargs)
        return self._backend

    def move(self, page: Any, to: Point) -> None:
        to = (float(to[0]), float(to[1]))
        if self._down:
            points, timings = trajectory(self.at, to, self._frequency)
            self._touch(page).move([(x, y, timings[i] - (timings[i - 1] if i else 0.0))
                                    for i, (x, y) in enumerate(points)])
        self.at = to

    def press(self, page: Any) -> None:
        self._touch(page).down(*self.at)
        self._down = True

    def release(self, page: Any) -> None:
        self._touch(page).up(*self.at)
        self._down = False

    def click(self, page: Any, to: Point) -> None:
        """A tap whose contact patch wobbles a pixel while held; a motionless tap is a synthetic one."""
        self.move(page, to)
        self.press(page)
        held = self._pause_ms("tap")
        _delay(held * 0.5)
        try:
            self._touch(page).move([(self.at[0] + random.gauss(0.0, 0.9), self.at[1] + random.gauss(0.0, 0.9), 0.0)])
        except Exception:
            pass
        _delay(held * 0.5)
        self.release(page)

    def type_text(self, page: Any, field: Any, text: str) -> bool:
        for clear, arg in (("clear", ()), ("fill", ("",))):
            fn = getattr(field, clear, None)
            if fn is not None:
                try:
                    fn(*arg)
                    break
                except Exception:
                    continue
        send = getattr(field, "send_keys", None)
        for ch in text:
            try:
                send(ch) if send is not None else page.keyboard.type(ch)
            except Exception as exc:
                _log(f"could not type into the captcha field: {exc}")
                return False
            self.pause("key")
        return True


class NullHumanizer(Humanizer):
    """No humanisation: one move per gesture and a single fill. Fast, and detectable."""

    name = "none"
    hovers = False

    def move(self, page: Any, to: Point) -> None:
        if _same_point(self.at, to):
            return
        self.at = (float(to[0]), float(to[1]))
        try:
            page.mouse.move(*self.at)
        except Exception as exc:
            if "Target closed" in str(exc) or "Session closed" in str(exc):
                _log("could not move mouse; page or session closed")

    def press(self, page: Any) -> None:
        page.mouse.down()

    def release(self, page: Any) -> None:
        page.mouse.up()

    def type_text(self, page: Any, field: Any, text: str) -> bool:
        try:
            field.fill(text)
            return True
        except Exception:
            pass
        try:
            page.keyboard.press("Control+A")
        except Exception:
            pass
        try:
            page.keyboard.type(text)
            return True
        except Exception as exc:
            _log(f"could not type into the captcha field: {exc}")
            return False


MODES = {"mouse": MouseHumanizer, "mobile": MobileHumanizer, "none": NullHumanizer}


def resolve(config: Any) -> Humanizer:
    """`config.humanizer`, else `config.humanization`, else CAPTCHA_HUMANIZATION, else mouse."""
    custom = getattr(config, "humanizer", None)
    if custom is not None:
        return custom
    mode = str(getattr(config, "humanization", None) or os.getenv("CAPTCHA_HUMANIZATION") or "mouse").strip().lower()
    if mode not in MODES:
        raise ValueError(f"unknown humanization mode {mode!r}; expected one of {', '.join(sorted(MODES))}, "
                         "or pass your own object as PageSolverConfig.humanizer")
    start = getattr(config, "starting_mouse_position", None) or (0.0, 0.0)
    if mode == "mobile":
        return MobileHumanizer(start, driver=getattr(config, "touch_driver", None),
                               **(getattr(config, "touch_transform", None) or {}))
    return MODES[mode](start)

File: js/src/humanize.ts
/**
 * How the driver moves: one pluggable object per input device. Mirrors python/humanize.py.
 *
 * The pointer position lives here, not in the solver, because a touch mode that dispatches no
 * motion between taps still has to say where the next gesture starts.
 */

import type { Page, PlaywrightElementHandle } from './playwright-types.js';
import { generateTrajectory } from 'cursory-js';

export type Point = [number, number];

const delay = (ms: number) => new Promise<void>((r) => setTimeout(r, Math.max(0, ms)));
const log = (m: string) => console.log(`[captchakraken] ${m}`);
const uniform = (lo: number, hi: number) => lo + Math.random() * (hi - lo);
const samePoint = (a: Point, b: Point) => Math.abs(a[0] - b[0]) < 1e-6 && Math.abs(a[1] - b[1]) < 1e-6;
const gauss = (sigma: number) => {
  let u = 0;
  while (u === 0) u = Math.random();
  return sigma * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * Math.random());
};
const closedError = (e: unknown) => /Target closed|Session closed/.test(e instanceof Error ? e.message : String(e));

/** A recorded human movement morphed onto the endpoints; timings are cumulative ms. */
export function trajectory(start: Point, end: Point, frequency: number): [Point[], number[]] {
  const { points, timings } = generateTrajectory(start, end, { frequency });
  return [points.map((q) => [Number(q[0]), Number(q[1])] as Point), timings.map(Number)];
}

/** Every inter-gesture wait the driver takes, named, so each device supplies its own table. */
export type PauseKind = 'tap' | 'between' | 'grab' | 'drop' | 'probe' | 'settle' | 'key';
export const PAUSE_KINDS: readonly PauseKind[] = ['tap', 'between', 'grab', 'drop', 'probe', 'settle', 'key'] as const;

export interface Humanizer {
  readonly name: string;
  /** False on a device with no resting cursor; switches off every hover-for-realism behaviour. */
  readonly hovers: boolean;
  at: Point;
  reset(page: Page): Promise<void>;
  move(page: Page, to: Point): Promise<void>;
  press(page: Page): Promise<void>;
  release(page: Page): Promise<void>;
  click(page: Page, to: Point): Promise<void>;
  drag(page: Page, src: Point, dst: Point): Promise<void>;
  typeText(page: Page, field: PlaywrightElementHandle, text: string): Promise<boolean>;
  pause(kind: PauseKind): Promise<void>;
}

type PauseTable = Record<string, [number, number]>;

export abstract class BaseHumanizer implements Humanizer {
  abstract readonly name: string;
  abstract readonly hovers: boolean;
  protected abstract readonly pauses: PauseTable;
  at: Point;

  constructor(start: Point = [0, 0]) {
    this.at = [Number(start[0]), Number(start[1])];
  }

  async reset(_page: Page): Promise<void> {}
  abstract move(page: Page, to: Point): Promise<void>;
  abstract press(page: Page): Promise<void>;
  abstract release(page: Page): Promise<void>;
  abstract typeText(page: Page, field: PlaywrightElementHandle, text: string): Promise<boolean>;

  protected pauseMs(kind: PauseKind): number {
    const range = this.pauses[kind];
    return range ? uniform(range[0], range[1]) : 0;
  }

  async pause(kind: PauseKind): Promise<void> {
    await delay(this.pauseMs(kind));
  }

  async click(page: Page, to: Point): Promise<void> {
    await this.move(page, to);
    await this.press(page);
    await this.pause('tap');
    await this.release(page);
  }

  async drag(page: Page, src: Point, dst: Point): Promise<void> {
    await this.move(page, src);
    await this.press(page);
    await this.pause('grab');
    await this.move(page, dst);
    await this.pause('drop');
    await this.release(page);
  }
}

export class MouseHumanizer extends BaseHumanizer {
  readonly name = 'mouse';
  readonly hovers = true;
  protected readonly pauses: PauseTable = {
    tap: [20, 50], between: [80, 160], grab: [50, 100], drop: [50, 100],
    probe: [40, 80], settle: [90, 210], key: [45, 135],
  };
  private frequency: number;

  constructor(start: Point = [0, 0], frequency = 60) {
    super(start);
    this.frequency = frequency;
  }

  async move(page: Page, to: Point): Promise<void> {
    if (samePoint(this.at, to)) return;
    const [points, timings] = trajectory(this.at, to, this.frequency);
    let viewport = { width: 1920, height: 1080 };
    try {
      viewport = page.viewportSize() ?? viewport;
    } catch { /* keep the default */ }
    const startTime = Date.now();
    for (let i = 0; i < points.length; i++) {
      const cx = Math.max(0, Math.min(points[i][0], viewport.width));
      const cy = Math.max(0, Math.min(points[i][1], viewport.height));
      try {
        await page.mouse.move(cx, cy);
      } catch (e) {
        if (closedError(e)) { log('could not move mouse; page or session closed'); return; }
        continue;
      }
      this.at = [cx, cy];
      if (timings[i] !== undefined) await delay(startTime + timings[i] - Date.now());
    }
  }

  async press(page: Page): Promise<void> { await page.mouse.down(); }
  async release(page: Page): Promise<void> { await page.mouse.up(); }

  async typeText(page: Page, _field: PlaywrightElementHandle, text: string): Promise<boolean> {
    try { await page.keyboard.press('Control+A'); } catch { /* the type below still replaces on most */ }
    for (const ch of text) {
      try {
        await page.keyboard.type(ch);
      } catch (e) {
        log(`could not type into the captcha field: ${e}`);
        return false;
      }
      await this.pause('key');
    }
    return true;
  }
}

/** `[x, y, dtMs]`: dt is the wait before that sample. */
export type TouchSample = [number, number, number];

/** Where touch events go. `move` takes a whole leg so a device driver can pace it itself. */
export interface TouchBackend {
  readonly name: string;
  down(x: number, y: number): Promise<void>;
  move(path: TouchSample[]): Promise<void>;
  up(x: number, y: number): Promise<void>;
}

/** `Input.dispatchTouchEvent` over CDP; Chromium-family pages launched with `hasTouch: true`. */
export class CdpTouchBackend implements TouchBackend {
  readonly name = 'cdp';
  private constructor(private session: any) {}

  static async open(page: Page): Promise<CdpTouchBackend> {
    try {
      return new CdpTouchBackend(await (page as any).context().newCDPSession(page));
    } catch (e) {
      throw new Error(
        `mobile humanisation needs touch dispatch, and this page offers none (${e}). ` +
        'Use a Chromium-family Playwright browser launched with hasTouch: true, ' +
        'or pass an Appium driver as CaptchaKrakenConfig.touchDriver.');
    }
  }

  private async send(kind: string, points: Array<[number, number]>): Promise<void> {
    // radius and force are what a real digitizer reports; a synthetic tap reports zero.
    await this.session.send('Input.dispatchTouchEvent', {
      type: kind,
      touchPoints: points.map(([x, y]) => ({
        x, y, radiusX: uniform(8, 14), radiusY: uniform(8, 14), force: uniform(0.35, 0.75), id: 1,
      })),
    });
  }

  async down(x: number, y: number): Promise<void> { await this.send('touchStart', [[x, y]]); }
  async move(path: TouchSample[]): Promise<void> {
    for (const [x, y, dt] of path) {
      if (dt > 0) await delay(dt);
      await this.send('touchMove', [[x, y]]);
    }
  }
  async up(_x: number, _y: number): Promise<void> { await this.send('touchEnd', []); }
}

/** CSS-pixel to device-pixel transform for a real handset. */
export interface TouchTransform {
  scale?: number;
  origin?: Point;
}

/** W3C touch pointer actions for Appium and WebdriverIO, paced by the device from one chain per leg. */
export class AppiumTouchBackend implements TouchBackend {
  readonly name = 'appium';
  private scale: number;
  private origin: Point;
  private scaleGiven: boolean;
  private checked = false;

  constructor(private driver: any, transform: TouchTransform = {}, private page?: Page) {
    this.scaleGiven = transform.scale !== undefined;
    this.scale = transform.scale ?? 1;
    this.origin = transform.origin ?? [0, 0];
  }

  /** Refuse an unset scale on a device whose pixel ratio is not 1: the finger would land elsewhere. */
  private async checkScale(): Promise<void> {
    this.checked = true;
    if (this.scaleGiven || !this.page) return;
    let dpr: number;
    try {
      dpr = Number(await (this.page as any).evaluate('() => window.devicePixelRatio'));
    } catch { return; }
    if (!Number.isFinite(dpr) || Math.abs(dpr - 1) < 1e-6) return;
    throw new Error(
      `the touch driver maps CSS pixels onto a device reporting devicePixelRatio ${dpr}, and no scale ` +
      `was given. Fix: touchTransform: { scale: ${dpr}, origin: [x, y] }, where origin is the webview's ` +
      'top-left in SCREEN coordinates, or pass scale: 1 to assert the coordinates are already mapped.');
  }

  private map(x: number, y: number): [number, number] {
    return [Math.round(this.origin[0] + x * this.scale), Math.round(this.origin[1] + y * this.scale)];
  }

  private async perform(actions: any[]): Promise<void> {
    const chain = [{ type: 'pointer', id: 'ck-finger', parameters: { pointerType: 'touch' }, actions }];
    if (typeof this.driver.performActions === 'function') return this.driver.performActions(chain);
    if (typeof this.driver.execute === 'function') return this.driver.execute('actions', { actions: chain });
    throw new Error("the touch driver speaks neither performActions() nor execute('actions', …); " +
      'pass a WebDriver-compatible driver or a custom TouchBackend.');
  }

  async down(x: number, y: number): Promise<void> {
    if (!this.checked) await this.checkScale();
    const [mx, my] = this.map(x, y);
    await this.perform([
      { type: 'pointerMove', duration: 0, origin: 'viewport', x: mx, y: my },
      { type: 'pointerDown', button: 0 },
    ]);
  }

  async move(path: TouchSample[]): Promise<void> {
    if (!this.checked) await this.checkScale();
    if (!path.length) return;
    await this.perform(path.map(([x, y, dt]) => {
      const [mx, my] = this.map(x, y);
      return { type: 'pointerMove', duration: Math.max(0, Math.round(dt)), origin: 'viewport', x: mx, y: my };
    }));
  }

  async up(_x: number, _y: number): Promise<void> {
    if (!this.checked) await this.checkScale();
    await this.perform([{ type: 'pointerUp', button: 0 }]);
  }
}

/** Playwright's tap-only touchscreen, for browsers with touch but no CDP. A drag cannot be expressed. */
export class TouchscreenTouchBackend implements TouchBackend {
  readonly name = 'touchscreen';
  private pending: Point | null = null;
  constructor(private page: any) {}

  async down(x: number, y: number): Promise<void> { this.pending = [x, y]; }
  async move(_path: TouchSample[]): Promise<void> {
    throw new Error('this browser exposes taps but not touch travel, so a drag/slide puzzle cannot be ' +
      'driven on it. Use a Chromium-family browser (CDP touch dispatch) or an Appium driver.');
  }
  async up(x: number, y: number): Promise<void> {
    const at = this.pending ?? [x, y];
    this.pending = null;
    await this.page.touchscreen.tap(at[0], at[1]);
  }
}

export async function touchBackendFor(page: Page, driver?: any, transform?: TouchTransform): Promise<TouchBackend> {
  if (driver) {
    if (typeof driver.down === 'function' && typeof driver.up === 'function') return driver as TouchBackend;
    return new AppiumTouchBackend(driver, transform, page);
  }
  try {
    return await CdpTouchBackend.open(page);
  } catch (e) {
    if ((page as any).touchscreen) {
      log('no CDP session; falling back to tap-only touch dispatch');
      return new TouchscreenTouchBackend(page);
    }
    throw e;
  }
}

/** A finger on glass: no hover, touch events only, slower and more variable pauses. */
export class MobileHumanizer extends BaseHumanizer {
  readonly name = 'mobile';
  readonly hovers = false;
  protected readonly pauses: PauseTable = {
    tap: [55, 130], between: [140, 320], grab: [90, 190], drop: [80, 170],
    probe: [70, 140], settle: [140, 300], key: [110, 320],
  };
  private backend: TouchBackend | null;
  private driver: any;
  private transform: TouchTransform | undefined;
  private frequency: number;
  private down = false;

  constructor(start: Point = [0, 0],
              options: { backend?: TouchBackend; driver?: any; transform?: TouchTransform; frequency?: number } = {}) {
    super(start);
    this.backend = options.backend ?? null;
    this.driver = options.driver;
    this.transform = options.transform;
    this.frequency = options.frequency ?? 90;
  }

  async reset(page: Page): Promise<void> {
    if (!this.down) return;
    try { await (await this.touch(page)).up(this.at[0], this.at[1]); } catch { /* a stale pointer must not fail a solve */ }
    this.down = false;
  }

  private async touch(page: Page): Promise<TouchBackend> {
    if (!this.backend) this.backend = await touchBackendFor(page, this.driver, this.transform);
    return this.backend;
  }

  async move(page: Page, to: Point): Promise<void> {
    const dest: Point = [Number(to[0]), Number(to[1])];
    if (this.down) {
      const [points, timings] = trajectory(this.at, dest, this.frequency);
      await (await this.touch(page)).move(points.map(([x, y], i) => [x, y, timings[i] - (i ? timings[i - 1] : 0)]));
    }
    this.at = dest;
  }

  async press(page: Page): Promise<void> {
    await (await this.touch(page)).down(this.at[0], this.at[1]);
    this.down = true;
  }

  async release(page: Page): Promise<void> {
    await (await this.touch(page)).up(this.at[0], this.at[1]);
    this.down = false;
  }

  /** A tap whose contact patch wobbles a pixel while held; a motionless tap is a synthetic one. */
  async click(page: Page, to: Point): Promise<void> {
    await this.move(page, to);
    await this.press(page);
    const held = this.pauseMs('tap');
    await delay(held / 2);
    try {
      await (await this.touch(page)).move([[this.at[0] + gauss(0.9), this.at[1] + gauss(0.9), 0]]);
    } catch { /* a tap-only backend cannot wobble; the tap still lands */ }
    await delay(held / 2);
    await this.release(page);
  }

  async typeText(page: Page, field: PlaywrightElementHandle, text: string): Promise<boolean> {
    const f = field as any;
    for (const [name, arg] of [['clear', undefined], ['fill', '']] as const) {
      if (typeof f?.[name] !== 'function') continue;
      try { await f[name](arg); break; } catch { /* an uncleared box is recoverable */ }
    }
    for (const ch of text) {
      try {
        if (typeof f?.sendKeys === 'function') await f.sendKeys(ch);
        else await page.keyboard.type(ch);
      } catch (e) {
        log(`could not type into the captcha field: ${e}`);
        return false;
      }
      await this.pause('key');
    }
    return true;
  }
}

/** No humanisation: one move per gesture and a single fill. Fast, and detectable. */
export class NullHumanizer extends BaseHumanizer {
  readonly name = 'none';
  readonly hovers = false;
  protected readonly pauses: PauseTable = {};

  async move(page: Page, to: Point): Promise<void> {
    if (samePoint(this.at, to)) return;
    this.at = [Number(to[0]), Number(to[1])];
    try {
      await page.mouse.move(this.at[0], this.at[1]);
    } catch (e) {
      if (closedError(e)) log('could not move mouse; page or session closed');
    }
  }

  async press(page: Page): Promise<void> { await page.mouse.down(); }
  async release(page: Page): Promise<void> { await page.mouse.up(); }

  async typeText(page: Page, field: PlaywrightElementHandle, text: string): Promise<boolean> {
    try { await (field as any).fill(text); return true; } catch { /* fall through to the keyboard */ }
    try { await page.keyboard.press('Control+A'); } catch { /* optional */ }
    try {
      await page.keyboard.type(text);
      return true;
    } catch (e) {
      log(`could not type into the captcha field: ${e}`);
      return false;
    }
  }
}

export type HumanizationMode = 'mouse' | 'mobile' | 'none';
export const MODES: readonly HumanizationMode[] = ['mouse', 'mobile', 'none'] as const;

export interface HumanizerOptions {
  humanization?: HumanizationMode;
  humanizer?: Humanizer;
  touchDriver?: any;
  touchTransform?: TouchTransform;
  startingMousePosition?: { x: number; y: number };
}

/** `config.humanizer`, else `config.humanization`, else CAPTCHA_HUMANIZATION, else mouse. */
export function resolveHumanizer(config: HumanizerOptions = {}): Humanizer {
  if (config.humanizer) return config.humanizer;
  const raw = (config.humanization ?? process.env.CAPTCHA_HUMANIZATION ?? 'mouse').toString().trim().toLowerCase() as HumanizationMode;
  if (!MODES.includes(raw)) {
    throw new Error(`unknown humanization mode '${raw}'; expected one of ${MODES.join(', ')}, ` +
      'or pass your own object as CaptchaKrakenConfig.humanizer');
  }
  const p = config.startingMousePosition;
  const start: Point = p ? [p.x, p.y] : [0, 0];
  if (raw === 'mobile') return new MobileHumanizer(start, { driver: config.touchDriver, transform: config.touchTransform });
  if (raw === 'none') return new NullHumanizer(start);
  return new MouseHumanizer(start);
}
